File: MetaheuristicOptimization/AllMetaheuristic/Nature-inspired/swarm-based/Joint Operation Algorithm_re/JOA/JOA.m
% Joint Operations Algorithm (JOA) Implementation
% Parameters
num_subpopulations = 5;      % Number of subpopulations
population_size_per_subpop = 10; % Population size per subpopulation
num_variables = 10;          % Number of variables
max_iterations = 100;        % Maximum number of iterations
mutation_rate = 0.1;         % Mutation rate
lower_limit = -5;            % Lower limit for variables
upper_limit = 5;             % Upper limit for variables

% Initialization
populations = cell(num_subpopulations, 1);  % Initialize populations
fitness = zeros(num_subpopulations, population_size_per_subpop);  % Initialize fitness values

% Generate random initial populations
for i = 1:num_subpopulations
    populations{i} = lower_limit + (upper_limit - lower_limit) * rand(population_size_per_subpop, num_variables);
end

% Main loop
for iter = 1:max_iterations
    % Evaluate fitness for each individual in each subpopulation
    for i = 1:num_subpopulations
        for j = 1:population_size_per_subpop
            fitness(i, j) = objective_function(populations{i}(j, :));
        end
    end
    
    % Update each subpopulation's position
    for i = 1:num_subpopulations
        for j = 1:population_size_per_subpop
            % Select a random subpopulation (excluding the current one)
            other_subpop_index = randi([1, num_subpopulations - 1]);
            if other_subpop_index >= i
                other_subpop_index = other_subpop_index + 1;
            end
            
            % Select a random individual from the selected subpopulation
            other_individual_index = randi([1, population_size_per_subpop]);
            
            % Move towards the selected individual
            direction = populations{other_subpop_index}(other_individual_index, :) - populations{i}(j, :);
            populations{i}(j, :) = populations{i}(j, :) + mutation_rate * direction;
            
            % Ensure the new position is within bounds
            populations{i}(j, :) = max(min(populations{i}(j, :), upper_limit), lower_limit);
        end
    end
end

% Find the best solution found among all subpopulations
all_fitness = fitness(:);
[best_fitness, best_index] = min(all_fitness);
best_solution_subpop = ceil(best_index / population_size_per_subpop);
best_solution_ind = mod(best_index - 1, population_size_per_subpop) + 1;
best_solution = populations{best_solution_subpop}(best_solution_ind, :);

% Display best solution found
fprintf('Best solution: %s\n', mat2str(best_solution));
fprintf('Best fitness: %f\n', best_fitness);
% Objective function (example)
function z = objective_function(x)
    % Example: Sphere function
    z = sum(x.^2);
end

File: chekml/MetaheuristicOptimization/CIntegration_ctypes/JOA.c
#include "JOA.h"
#include "generaloptimizer.h"
#include <stdlib.h>
#include <time.h>

// Function to generate a random double between min and max
double rand_double(double min, double max);

// Initialize subpopulations
void initialize_subpopulations(Optimizer *opt) {
    opt->population_size = NUM_SUBPOPULATIONS * POPULATION_SIZE_PER_SUBPOP;
    opt->population = (Solution *)malloc(opt->population_size * sizeof(Solution));
    if (!opt->population) {
        fprintf(stderr, "Memory allocation failed for population\n");
        exit(1);
    }

    for (int i = 0; i < opt->population_size; i++) {
        opt->population[i].position = (double *)malloc(opt->dim * sizeof(double));
        if (!opt->population[i].position) {
            fprintf(stderr, "Memory allocation failed for individual %d\n", i);
            exit(1);
        }

        // Randomly initialize position within bounds
        for (int k = 0; k < opt->dim; k++) {
            opt->population[i].position[k] = rand_double(opt->bounds[2 * k], opt->bounds[2 * k + 1]);
        }
        opt->population[i].fitness = INFINITY;
    }

    // Set optimize function pointer
    opt->optimize = JOA_optimize;
}

// Evaluate fitness for all subpopulations
void evaluate_subpopulations(Optimizer *opt, double (*objective_function)(double *)) {
    for (int i = 0; i < NUM_SUBPOPULATIONS; i++) {
        int start_idx = i * POPULATION_SIZE_PER_SUBPOP;
        for (int j = 0; j < POPULATION_SIZE_PER_SUBPOP; j++) {
            int idx = start_idx + j;
            double fitness = objective_function(opt->population[idx].position);
            opt->population[idx].fitness = fitness;

            // Update best solution
            if (fitness < opt->best_solution.fitness) {
                opt->best_solution.fitness = fitness;
                for (int k = 0; k < opt->dim; k++) {
                    opt->best_solution.position[k] = opt->population[idx].position[k];
                }
            }
        }
    }
}

// Update subpopulations
void update_subpopulations(Optimizer *opt) {
    for (int i = 0; i < NUM_SUBPOPULATIONS; i++) {
        int start_idx = i * POPULATION_SIZE_PER_SUBPOP;
        for (int j = 0; j < POPULATION_SIZE_PER_SUBPOP; j++) {
            int idx = start_idx + j;

            // Select a random subpopulation (excluding the current one)
            int other_subpop_idx = rand() % (NUM_SUBPOPULATIONS - 1);
            if (other_subpop_idx >= i) {
                other_subpop_idx++;
            }

            // Select a random individual from the other subpopulation
            int other_start_idx = other_subpop_idx * POPULATION_SIZE_PER_SUBPOP;
            int other_ind_idx = other_start_idx + (rand() % POPULATION_SIZE_PER_SUBPOP);

            // Move towards the selected individual
            for (int k = 0; k < opt->dim; k++) {
                double direction = opt->population[other_ind_idx].position[k] - 
                                  opt->population[idx].position[k];
                opt->population[idx].position[k] += MUTATION_RATE * direction;
            }
        }
    }

    // Enforce bounds
    enforce_bound_constraints(opt);
}

// Main optimization function
void JOA_optimize(void *opt_void, ObjectiveFunction objective_function) {
    Optimizer *opt = (Optimizer *)opt_void;
    srand(time(NULL)); // Seed random number generator

    initialize_subpopulations(opt);

    for (int iter = 0; iter < opt->max_iter; iter++) {
        evaluate_subpopulations(opt, objective_function);
        update_subpopulations(opt);

        // Debugging output
        printf("Iteration %d: Best Value = %f\n", iter + 1, opt->best_solution.fitness);
    }

    // Clean up population
    for (int i = 0; i < opt->population_size; i++) {
        free(opt->population[i].position);
    }
    free(opt->population);
    opt->population = NULL;
}
